packages/toolpad-app/src/server/withReqResLogs.ts

```diff
@@ -0,0 +1,87 @@
+import { NextApiHandler, NextApiRequest, NextApiResponse } from 'next';
+import * as _ from 'lodash-es';
+import { uuidv4 } from '../utils/uuid';
+import logger from './logger';
+
+function getLogMessageInfo(req: NextApiRequest) {
+  const { type, name } = req.body;
+  return type && name ? `${type}:${name}` : '';
+}
+
+function getBaseLogProperties(req: NextApiRequest) {
+  return {
+    type: req.body.type,
+    name: req.body.name,
+    ipAddress: req.headers['x-forwarded-for'] || req.socket?.remoteAddress,
+    host: req.headers.host,
+    userAgent: req.headers['user-agent'],
+    timestamp: new Date().getTime(),
+  };
+}
+
+function logRequestResponse(req: NextApiRequest, res: NextApiResponse, requestId: string) {
+  const oldWrite = res.write;
+  const oldEnd = res.end;
+
+  const chunks: Buffer[] = [];
+
+  res.write = (...restArgs: any[]) => {
+    chunks.push(Buffer.from(restArgs[0]));
+    return oldWrite.apply(
+      res,
+      restArgs as [
```

Comment:
Would it be useful to define types outside the methods? (not inline)

Reply:
this part was very hard to type because some of those functions have different call signatures - i'll see if i can remove that duplication, there's a slight difference there, not sure if it's the reason why i didn't reuse the type. i'll give it another look!